OpenRA.Game/LocalPlayerProfile.cs

```diff
@@ -0,0 +1,185 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using OpenRA.Network;
+
+namespace OpenRA
+{
+	public sealed class LocalPlayerProfile
+	{
+		const int AuthKeySize = 2048;
+		public enum LinkState { Uninitialized, GeneratingKeys, Unlinked, CheckingLink, ConnectionFailed, Linked }
+
+		public LinkState State { get { return innerState; } }
+		public string Fingerprint { get { return innerFingerprint; } }
+		public string PublicKey { get { return innerPublicKey; } }
+
+		public PlayerProfile ProfileData { get { return innerData; } }
+
+		volatile LinkState innerState;
+		volatile PlayerProfile innerData;
+		volatile string innerFingerprint;
+		volatile string innerPublicKey;
+
+		RSAParameters parameters;
+		readonly string filePath;
+		readonly PlayerDatabase playerDatabase;
+
+		public LocalPlayerProfile(string filePath, PlayerDatabase playerDatabase)
+		{
+			this.filePath = filePath;
+			this.playerDatabase = playerDatabase;
+			innerState = LinkState.Uninitialized;
+
+			try
+			{
+				if (File.Exists(filePath))
+				{
+					using (var rsa = new RSACryptoServiceProvider())
+					{
+						using (var data = File.OpenRead(filePath))
+						{
+							var keyData = Convert.FromBase64String(data.ReadAllText());
+							rsa.FromXmlString(new string(Encoding.ASCII.GetChars(keyData)));
+						}
+
+						parameters = rsa.ExportParameters(true);
+						innerPublicKey = CryptoUtil.EncodePEMPublicKey(parameters);
+						innerFingerprint = CryptoUtil.SHA1Hash(innerPublicKey);
```

Comment:
This is crap.  I should look into hashing `parameters` directly here, and then implementing something similar from the decoded key on the backend.